ansible-base seems not working

**Describe the bug**

ansible-base seems not working

**Steps to reproduce**

❯ asdf plugin add ansible-base https://github.com/amrox/asdf-pyapp.git

❯ asdf current
ansible-base    ______          No version is set. Run "asdf <global|shell|local> ansible-base <version>"
awscli          2.4.0           /root/.tool-versions
packer          1.7.8           /root/.tool-versions
terraform       1.0.11          /root/.tool-versions

❯ asdf list all ansible-base
Plugin ansible-base's list-all callback script failed with output:


❯ asdf install ansible-base latest
Plugin ansible-base's list-all callback script failed with output:


ansible-base  is already installed

re-hello @amrox 

In fact I have a similar (but not exactly the same) issue on another Linux (ArchLinux).
``` bash
# uname -a && cat /etc/os-release
Linux 99348e6ad5c6 5.4.139-16311-g330b3df0010b #1 SMP PREEMPT Mon Nov 8 12:23:05 PST 2021 x86_64 GNU/Linux
NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://bugs.archlinux.org/"
LOGO=archlinux-logo
```

``` bash
# asdf current
ansible-base    ______          No version set. Run "asdf <global|shell|local> ansible-base <version>"
```

When I try to install `latest` ansible I have this output :
``` bash
# asdf install ansible-base latest
asdf-pyapp: [ERROR] Unable to parse versions for 'ansible-base'
ansible-base  is already installed
```
But if I use a specific version, it work

``` bash
asdf install ansible-base 2.10.15
Requirement already satisfied: pip in ./.asdf/installs/ansible-base/2.10.15/venv/lib/python3.9/site-packages (21.2.4)
Collecting pip
  Downloading pip-21.3.1-py3-none-any.whl (1.7 MB)
     |████████████████████████████████| 1.7 MB 4.1 MB/s 
Collecting wheel
  Downloading wheel-0.37.0-py2.py3-none-any.whl (35 kB)
Installing collected packages: wheel, pip
  Attempting uninstall: pip
    Found existing installation: pip 21.2.4
    Uninstalling pip-21.2.4:
      Successfully uninstalled pip-21.2.4
Successfully installed pip-21.3.1 wheel-0.37.0
Collecting ansible-base==2.10.15
  Downloading ansible-base-2.10.15.tar.gz (6.1 MB)

[...]

Installing collected packages: pyparsing, click, userpath, packaging, argcomplete, pipx
Successfully installed argcomplete-1.12.3 click-8.0.3 packaging-21.3 pipx-0.16.3 pyparsing-3.0.6 userpath-1.7.0
```

``` bash
# asdf global ansible-base  2.10.15

# ansible --version
ansible 2.10.15
  config file = None
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /root/.asdf/installs/ansible-base/2.10.15/venv/lib/python3.9/site-packages/ansible
  executable location = /root/.asdf/installs/ansible-base/2.10.15/bin/ansible
  python version = 3.9.9 (main, Nov 20 2021, 21:30:06) [GCC 11.1.0]
```

So this issue occur only with  `latest` keyword

--- a/lib/utils.bash
+++ b/lib/utils.bash
@@ -111,15 +111,19 @@ get_package_versions() {
 
   local pip_version
   pip_version=$(get_python_pip_versions "$ASDF_PYAPP_RESOLVED_PYTHON_PATH")
-  if [[ $pip_version =~ ^([0-9]+)\. ]]; then
+  if [[ $pip_version =~ ^([0-9]+)\.([0-9]+)\. ]]; then
     local pip_version_major=${BASH_REMATCH[1]}
+    local pip_version_minor=${BASH_REMATCH[2]}
   else
-    fail "Unable to parse pip major version"
+    fail "Unable to parse pip version"
   fi
 
   local pip_install_args=()
   local version_output_raw
-  if [ "${pip_version_major}" -ge 20 ]; then
+
+  # we rely on the "legacy resolver" to get versions, which was introduced in 20.3
+  if [ "${pip_version_major}" -ge 21 ] ||
+    { [ "${pip_version_major}" -eq 20 ] && [ "${pip_version_minor}" -ge 3 ]; }; then
     pip_install_args+=("--use-deprecated=legacy-resolver")
   fi
   version_output_raw=$("${ASDF_PYAPP_RESOLVED_PYTHON_PATH}" -m pip install ${pip_install_args[@]+"${pip_install_args[@]}"} "${package}==" 2>&1) || true

--- a/lib/utils.bash
+++ b/lib/utils.bash
@@ -119,7 +119,7 @@ get_package_versions() {
 
   local pip_install_args=()
   local version_output_raw
-  if [ "${pip_version_major}" -gt 20 ]; then
+  if [ "${pip_version_major}" -ge 20 ]; then
     pip_install_args+=("--use-deprecated=legacy-resolver")
   fi
   version_output_raw=$("${ASDF_PYAPP_RESOLVED_PYTHON_PATH}" -m pip install ${pip_install_args[@]+"${pip_install_args[@]}"} "${package}==" 2>&1) || true

--- a/test/integration-tests.bats
+++ b/test/integration-tests.bats
@@ -103,9 +103,99 @@ teardown() {
   assert_output --partial "Failed to find python3 >= 3.6"
 }
 
-@test "check latest with pip version 20" {
+@test "check latest with pip version 19.3" {
+
+  local pip_version=19.3.1
+
+  in_container asdf global python 3.8.10
+
+  in_container python3 -m pip install --upgrade pip=="$pip_version"
+
+  run in_container python3 -m pip --version
+  assert_output --partial "$pip_version"
+
+  in_container cp -r /root/asdf-pyapp /root/.asdf/plugins/cowsay
+  run in_container asdf list all cowsay
+  # TODO: "asdf list all" seems to mask return codes. It exits 0 even if we
+  # exit nonzero in our function. So just check output for errors for now...
+  refute_output --partial "asdf-pyapp: [ERROR]"
+}
+
+@test "check latest with pip version 20.0" {
+
+  local pip_version=20.0.1
+
+  in_container asdf global python 3.8.10
+
+  in_container python3 -m pip install --upgrade pip=="$pip_version"
+
+  run in_container python3 -m pip --version
+  assert_output --partial "$pip_version"
+
+  in_container cp -r /root/asdf-pyapp /root/.asdf/plugins/cowsay
+  run in_container asdf list all cowsay
+  # TODO: "asdf list all" seems to mask return codes. It exits 0 even if we
+  # exit nonzero in our function. So just check output for errors for now...
+  refute_output --partial "asdf-pyapp: [ERROR]"
+}
+
+@test "check latest with pip version 20.1" {
+
+  local pip_version=20.1.1
+
+  in_container asdf global python 3.8.10
+
+  in_container python3 -m pip install --upgrade pip=="$pip_version"
+
+  run in_container python3 -m pip --version
+  assert_output --partial "$pip_version"
+
+  in_container cp -r /root/asdf-pyapp /root/.asdf/plugins/cowsay
+  run in_container asdf list all cowsay
+  # TODO: "asdf list all" seems to mask return codes. It exits 0 even if we
+  # exit nonzero in our function. So just check output for errors for now...
+  refute_output --partial "asdf-pyapp: [ERROR]"
+}
+
+@test "check latest with pip version 20.2" {
+
+  local pip_version=20.2.4
+
+  in_container asdf global python 3.8.10
+
+  in_container python3 -m pip install --upgrade pip=="$pip_version"
+
+  run in_container python3 -m pip --version
+  assert_output --partial "$pip_version"
+
+  in_container cp -r /root/asdf-pyapp /root/.asdf/plugins/cowsay
+  run in_container asdf list all cowsay
+  # TODO: "asdf list all" seems to mask return codes. It exits 0 even if we
+  # exit nonzero in our function. So just check output for errors for now...
+  refute_output --partial "asdf-pyapp: [ERROR]"
+}
+
+@test "check latest with pip version 20.3" {
 
   local pip_version=20.3.4
+
+  in_container asdf global python 3.8.10
+
+  in_container python3 -m pip install --upgrade pip=="$pip_version"
+
+  run in_container python3 -m pip --version
+  assert_output --partial "$pip_version"
+
+  in_container cp -r /root/asdf-pyapp /root/.asdf/plugins/cowsay
+  run in_container asdf list all cowsay
+  # TODO: "asdf list all" seems to mask return codes. It exits 0 even if we
+  # exit nonzero in our function. So just check output for errors for now...
+  refute_output --partial "asdf-pyapp: [ERROR]"
+}
+
+@test "check latest with pip version 21.3" {
+
+  local pip_version=21.3.1
 
   in_container asdf global python 3.8.10
 

--- a/test/integration-tests.bats
+++ b/test/integration-tests.bats
@@ -1,5 +1,4 @@
-
-in_container () {
+in_container() {
   args=($*)
   docker exec -it "$CONTAINER" bash -l -c "${args[*]@Q}"
 }
@@ -8,7 +7,7 @@ setup() {
   load 'test_helper/bats-support/load'
   load 'test_helper/bats-assert/load'
 
-  MY_DIR="$( cd "$( dirname "$BATS_TEST_FILENAME" )" >/dev/null 2>&1 && pwd )"
+  MY_DIR="$(cd "$(dirname "$BATS_TEST_FILENAME")" >/dev/null 2>&1 && pwd)"
   SRCROOT=$(dirname "$MY_DIR")
 
   TAG="asdf-pyapp-integration-bionic"
@@ -32,7 +31,7 @@ teardown() {
   in_container asdf plugin remove python
 
   run in_container which python3
-  assert_output --partial /usr/bin/python3  #TODO: why is --partial required? newline?
+  assert_output --partial /usr/bin/python3 #TODO: why is --partial required? newline?
 
   in_container cp -r /root/asdf-pyapp /root/.asdf/plugins/cowsay
   in_container asdf install cowsay 4.0
@@ -72,7 +71,7 @@ teardown() {
   in_container cowsay "woo woo"
 
   run in_container readlink /root/.asdf/installs/cowsay/4.0/venv/bin/python3
-  assert_output --partial  /root/.asdf/installs/python/3.8.10/bin/python3
+  assert_output --partial /root/.asdf/installs/python/3.8.10/bin/python3
 }
 
 @test "install with asdf python 3.5.10 system python 3.6" {
@@ -102,6 +101,24 @@ teardown() {
   in_container cp -r /root/asdf-pyapp /root/.asdf/plugins/cowsay
   run in_container asdf install cowsay 4.0
   assert_output --partial "Failed to find python3 >= 3.6"
+}
+
+@test "check latest with pip version 20" {
+
+  local pip_version=20.3.4
+
+  in_container asdf global python 3.8.10
+
+  in_container python3 -m pip install --upgrade pip=="$pip_version"
+
+  run in_container python3 -m pip --version
+  assert_output --partial "$pip_version"
+
+  in_container cp -r /root/asdf-pyapp /root/.asdf/plugins/cowsay
+  run in_container asdf list all cowsay
+  # TODO: "asdf list all" seems to mask return codes. It exits 0 even if we
+  # exit nonzero in our function. So just check output for errors for now...
+  refute_output --partial "asdf-pyapp: [ERROR]"
 }
 
 @test "check \$ASDF_PYAPP_DEFAULT_PYTHON_PATH works" {
@@ -240,7 +257,7 @@ setup_direnv() {
 
   run in_container readlink /root/.asdf/installs/cowsay/4.0/venv/bin/python3
   #assert_output --partial /usr/bin/python3
-  assert_output --partial  /root/.asdf/installs/python/3.8.10/bin/python3
+  assert_output --partial /root/.asdf/installs/python/3.8.10/bin/python3
 }
 
 @test "check ASDF_PYAPP_VENV_COPY_MODE=1" {
@@ -261,7 +278,8 @@ setup_direnv() {
 check_app() {
   local app="$1"
   local version="$2"
-  shift; shift
+  shift
+  shift
 
   in_container asdf global python system
   in_container cp -r /root/asdf-pyapp /root/.asdf/plugins/"$app"

